From: Eric Dumazet <eric.dumazet@gmail.com>
Date: Thu, 13 Oct 2011 07:50:19 +0000
Subject: [PATCH 04/27] bnx2: fix skb truesize underestimation
Patch-mainline: v3.2-rc1
Git-commit: a1f4e8bcbccf50cf1894c263af4d677d4f566533
References: bnc#790590 FATE#313913

bnx2 allocates a full page per fragment. We must account PAGE_SIZE
increments on skb->truesize, not the actual frag length.

Signed-off-by: Eric Dumazet <eric.dumazet@gmail.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Michael Chan <mchan@broadcom.com>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/ethernet/broadcom/bnx2.c |    3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

--- a/drivers/net/ethernet/broadcom/bnx2.c
+++ b/drivers/net/ethernet/broadcom/bnx2.c
@@ -3054,7 +3054,6 @@ bnx2_rx_skb(struct bnx2 *bp, struct bnx2
 						&skb_shinfo(skb)->frags[i - 1];
 					frag->size -= tail;
 					skb->data_len -= tail;
-					skb->truesize -= tail;
 				}
 				return 0;
 			}
@@ -3086,7 +3085,7 @@ bnx2_rx_skb(struct bnx2 *bp, struct bnx2
 
 			frag_size -= frag_len;
 			skb->data_len += frag_len;
-			skb->truesize += frag_len;
+			skb->truesize += PAGE_SIZE;
 			skb->len += frag_len;
 
 			pg_prod = NEXT_RX_BD(pg_prod);
